// simple-xdm/src/common/util.js
const LOG_PREFIX = "[Simple-XDM] ";

class Util {

  locationOrigin() {
    if (!window.location.origin) {
      return window.location.protocol + "//" + window.location.hostname + ( window.location.port ? ':' + window.location.port : '' );
    } else {
      return window.location.origin;
    }
  }

  randomString() {
    return Math.floor(Math.random() * 1000000000).toString(16);
  }

  isString(str) {
    return typeof str === "string" || str instanceof String;
  }

  argumentsToArray(arrayLike) {
    return Array.prototype.slice.call(arrayLike);
  }

  argumentNames(fn) {
    return fn.toString()
      .replace(/((\/\/.*$)|(\/\*[^]*?\*\/))/mg, '') // strip comments
      .replace(/[^(]+\(([^)]*)[^]+/, '$1') // get signature
      .match(/([^\s,]+)/g) || [];
  }

  hasCallback(args) {
    var length = args.length;
    return length > 0 && typeof args[length - 1] === 'function';
  }

  error(msg) {
      if (window.console) {
          console.error(LOG_PREFIX + msg);
      }
  }

  warn(msg) {
      if (window.console) {
          console.warn(LOG_PREFIX + msg);
      }
  }

  _bind(thisp, fn){
    if(Function.prototype.bind) {
      return fn.bind(thisp);
    }
    return function () {
      return fn.apply(thisp, arguments);
    };
  }

  each(o, it) {
    var l;
    var k;
    if (o) {
      l = o.length;
      if (l != null && typeof o !== 'function') {
        k = 0;
        while (k < l) {
          if (it.call(o[k], k, o[k]) === false){
            break;
          }
          k += 1;
        }
      } else {
        for (k in o) {
          if (o.hasOwnProperty(k)) {
            if (it.call(o[k], k, o[k]) === false){
              break;
            }
          }
        }
      }
    }
  }

  extend(dest) {
    var args = arguments;
    var srcs = [].slice.call(args, 1, args.length);
    srcs.forEach(function(source) {
      if(typeof source === "object") {
        Object.getOwnPropertyNames(source).forEach(function (name) {
          dest[name] = source[name];
        });
      }
    });
    return dest;
  }

}

module.exports = new Util();


/** WEBPACK FOOTER **
 ** ./~/simple-xdm/src/common/util.js
 **/
